lib/phoenix/token.ex

```diff
@@ -0,0 +1,110 @@
+defmodule Phoenix.Token do
+  @moduledoc """
+  Tokens provide a way to  generate, verify, and decrypt bearer
+  tokens for use in Channels or API authentication.
+
+  ## Basic Usage
+  When generating a unique token for usage in an API or Channel
+  it is advised to use a unique identifier for the user typically
+  the id from a Database. For example:
+
+      iex> user_id = 1
+      iex> token = sign_token(endpoint, "user", user_id)
+      iex> user_id == verify_token(endpoint, token)
+      true
+
+  In that example we have a user's id, we generate a token and send
+  it to the client. When the client uses the token in a futher
+  communication you verify the token, query the database and authorize
+  the user.
+
+  When using it with a socket a typical example might be:
+
+    defmodule MyChannel do
+      def join("my:" <> id = topic, %{token: token}, socket) do
+        case verify_token(socket, "user", token) do
+          user_id ->
+            socket = assigns(socket, :user, Repo.get!(User, user_id))
+            reply socket, "join", %{}
+          :expired -> :ignore
+          :error -> :ignore
+        end
+      end
+    end
+
+  In this example the Phoenix.js client will be sending up the token
+  in the join command.
+
+  If you want to provide an API to the user send it down to the user
+  by generating it in a controller or view action.
+
+    def create(conn, params) do
+      user = User.create(params)
+      render conn, "user.json", %{token: sign_token(conn, "user", user.id), user: user}
+    end
+
+    Then the client can use it all the way through. If you'd like to
+    use the same token for API authorization then I suggest adding it
+    as a header to your authroized http requests
+
+      Authorization: Bearer #\{token}
+
+    Then create a plug to verify the token and authorize with the database.
+
+  """
+  alias Plug.Crypto.KeyGenerator
+  alias Plug.Crypto.MessageVerifier
+
+  @doc """
+  Signs your data into a token you can send down to clients
+  """
+  def sign_token(context, salt,  data, opts \\ []) when is_binary(data) or is_integer(data) or is_map(data) do
+    secret = get_endpoint(context) |> get_secret(salt)
+
+    max_age = if Dict.has_key?(opts, :max_age) do
+      opts[:max_age]
```

Comment:
So, I wouldn't use max_age when signing. Simply store the timestamp on sign. I would use the max_age when verifying. This is because if you accidentally set the max_age high, you can still change the verification and everything will be fine.